### PR TITLE
LPS-202532 Internal server error when a already existing ERC is used in a POST API call

### PR DESCRIPTION
**Pull request by magjed4289**
https://issues.liferay.com/browse/LPS-202532

### CODE REVIEW

**Comment by magjed4289** [comment]
ci:test:relevant

**Comment by magjed4289** [comment]
ci:test:stable

**Comment by magjed4289** [comment]
ci:test:stable

**Comment by magjed4289** [comment]
ci:sf

**Comment by magjed4289** [comment]
ci:stop

**Comment by magjed4289** [comment]
ci:test:relevant

**Comment by magjed4289** [comment]
ci:test:sf

**Review by 4lejandrito — APPROVED** [review]
@magjed4289 read my commits (the messages and the code). In my opinion the fix was in the right place and the tests are good! I just restored the original logic and added a better error message for the logs.

**Comment by 4lejandrito** [comment]
ci:test:relevant

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java [hunk 1]
```diff
@@ -4023,18 +4023,20 @@ private void _validateAutoIncrementValue(
...
 			externalReferenceCode, companyId, objectDefinitionId);
 
 		if ((objectEntry != null) &&
 			(objectEntry.getObjectEntryId() != objectEntryId)) {
 
-			throw new ObjectEntryValuesException.MustNotBeDuplicate(
-				externalReferenceCode);
+			throw new DuplicateObjectEntryExternalReferenceCodeException(
+				StringBundler.concat(
+					"Duplicate object entry with external reference code ",
```
Comment (magjed4289) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:4037]: @4lejandrito isn't the practice now not to put specific data in the message body for security reasons?

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java [hunk 2]
```diff
@@ -4024,17 +4023,14 @@ private void _validateAutoIncrementValue(
...
 
 	private void _validateExternalReferenceCode(
-		String externalReferenceCode, long companyId, long groupId,
-		long objectEntryId, boolean newObjectEntry) {
+		String externalReferenceCode, long companyId, long objectDefinitionId,
+		long objectEntryId) {
 
-		ObjectEntry ercObjectEntries = objectEntryPersistence.fetchByERC_G_C(
-			externalReferenceCode, groupId, companyId);
+		ObjectEntry objectEntry = objectEntryPersistence.fetchByERC_C_ODI(
```
Comment (magjed4289) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:4029]: @4lejandrito either this or not setting the groupId and companyId before makes some cases fail.
My first solution was to use both, fetchByERC_G_C and fetchByERC_C_ODI, however, leaving only fetchByERC_G_C worked for all the cases (but I understand that business rule of scoped by definitionId changes). I can keep investigate why the errors.

image

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java (hunk 1 above)
Comment (4lejandrito) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:4037]: This message is not exposed to the user, just to the logs. The message displayed to the user is [the one from the exception mapper](https://github.com/liferay/liferay-portal/blob/0bb9010ffb046dc55c0019a4a3307dbc894e8db2/modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/exception/mapper/DuplicateExternalReferenceCodeExceptionMapper.java#L32-L36).

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java (hunk 2 above)
Comment (4lejandrito) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:4030]: Hey @magjed4289 I just pushforced to move the call to `_setExternalReferenceCode` back down to make sure the objectDefinitionId is already populated (it was 0, that's why the right exception was not being thrown). With this change all the tests pass.

**Comment by 4lejandrito** [comment]
ci:test:relevant

**Comment by 4lejandrito** [comment]
ci:forward:force

**Comment by 4lejandrito** [comment]
ci:stop

**Comment by 4lejandrito** [comment]
https://github.com/brianchandotcom/liferay-portal/pull/144879

**Pull request closed without merging**